bootstrap/kubeadm/internal/cloudinit/controlplane_join.go

```diff
@@ -18,11 +18,136 @@ package cloudinit
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Log an error and exit.
+# Args:
+#   $1 Message to log with the error
+#   $2 The error code to return
+log::error_exit() {
+  local message="${1}"
+  local code="${2}"
+
+  log::error "${message}"
+  log::info "Removing member from cluster status"
+  kubeadm reset -f update-cluster-status || true
+  log::info "Removing etcd member"
+  kubeadm reset -f remove-etcd-member || true
+  log::info "Resetting kubeadm"
+  kubeadm reset -f || true
+  log::error "cluster.x-k8s.io kubeadm bootstrap script $0 exiting with status ${code}"
+  exit "${code}"
+}
+
+log::success_exit() {
+  log::info "cluster.x-k8s.io kubeadm bootstrap script $0 finished"
+  exit 0
+}
+
+# Log an error but keep going.
+log::error() {
+  local message="${1}"
+  timestamp=$(date --iso-8601=seconds)
+  echo "!!! [${timestamp}] ${1}" >&2
+  shift
+  for message; do
+    echo "    ${message}" >&2
+  done
+}
+
+# Print a status line.  Formatted to show up in a stream of output.
+log::info() {
+  timestamp=$(date --iso-8601=seconds)
+  echo "+++ [${timestamp}] ${1}"
+  shift
+  for message; do
+    echo "    ${message}"
+  done
+}
+
+check_kubeadm_command() {
+  local command="${1}"
+  local code="${2}"
+  local out="${3}"
+  local sanitised="${out//[$'\t\r\n']/}"
+  case ${code} in
+  "0")
+    log::info "kubeadm reported successful execution for ${command}"
+    ;;
+  "1")
+    log::error "kubeadm reported failed action(s) for ${command}"
+    log::error "${sanitised}"
+    ;;
+  "2")
+    log::error "kubeadm reported preflight check error during ${command}"
+    log::error "${sanitised}"
+    ;;
+  "3")
+    log::error "kubeadm reported validation error for ${command}"
+    log::error_exit "${sanitised}"
+    ;;
+  *)
+    log::error "kubeadm reported unknown error ${code} for ${command}"
+    log::error "${sanitised}"
+    ;;
+  esac
+}
+
+function retry-command() {
+  n=0
+  local kubeadm_return
+  until [ $n -ge 5 ]; do
+    local out
+    out=$(
+      "$@" --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
+    )
+    kubeadm_return=$?
+    check_kubeadm_command "'$*'" "${kubeadm_return}" "${out}"
+    if [ ${kubeadm_return} -eq 0 ]; then
+      break
+    fi
+    # We allow preflight errors to pass
+    if [ ${kubeadm_return} -eq 2 ]; then
+      break
+    fi
+    n=$((n + 1))
+    log::info
+    sleep 15
+  done
+  if [ ${kubeadm_return} -ne 0 ]; then
+    log::error_exit "too many errors, exiting"
+  fi
+}
+
+retry-command kubeadm join phase preflight
+retry-command kubeadm join phase control-plane-prepare download-certs
+retry-command kubeadm join phase control-plane-prepare certs
+retry-command kubeadm join phase control-plane-prepare kubeconfig
+retry-command kubeadm join phase kubelet-start
```

Comment:
By using phase we are loosing a check for etcd health before joining, see https://github.com/kubernetes/kubernetes/blob/c441a1a7dcca81aa6bd7f9b0f5babe6e8a590c7b/cmd/kubeadm/app/cmd/join.go#L207
This can make the join sequence less resilient to the etcd cluster startup slow-downs

Reply:
Can I invoke the health check whilst using phases?

Reply:
i don't remember why we hid this phase, maybe because it is reproducible outside of kubeadm.
it's the analog of calling `etcdctl cluster-health`, yet this opens the can of worms with the recent `etcdctl` certificate flag breaking changes.

given there is a retry on `retry-command kubeadm join phase control-plane-join etcd` we can say "hey, it's not checking the etcd cluster status, but at least it will retry the setup the local server instance."

Reply:
had a look through if it was trivial to replicate on the command line, but kubeadm is generating it's own etcd client and connecting in, so probably not. to be fair, Cluster API should be checking etcd state before we even get here, so should be fine?